File: examples/app/swiftui/ExamplesApp/ContentView.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import SwiftUI
import ExamplesAppKotlin

struct ContentView: View {
    init(todoList: TodoList) {
        self.todoList = todoList
    }

    var todoList: TodoList
    let stride: UInt64 = UInt64(1)
    @State private var clicked = UInt64.zero
    @State private var text = ""
    var body: some View {
        VStack {
            HStack {
                Text("\(clicked)").padding()
                TextField("New task", text: $text, onCommit:  {
                    try! self.todoList.addEntry(entry: TodoEntry(text: "\(clicked) \(text)"))
                    text = ""
                    clicked = try! Uniffi_kmm_example_arithmetic_procmacro_commonKt.add(a: clicked, b: stride)
                }).padding()
            }

            List{
                Section(header: Text("Still To Do")) {
                    ForEach(todoList.getEntries(), id: \.self) { entry in
                        Text(entry.text)
                    }
                    .onDelete { index in
                        if let i = index.first {
                            let entry = todoList.getEntries()[i]
                            try! todoList.clearItem(todo: entry.text)
                            clicked = try! Uniffi_kmm_example_arithmetic_procmacro_commonKt.sub(a: clicked, b: stride)
                        }
                    }
                }
            }
        }
    }
}

#Preview {
    ContentView(todoList: TodoList())
}

File: examples/app/src/androidMain/kotlin/ContentView.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package io.gitlab.trixnity.uniffi.examples.app

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import io.gitlab.trixnity.uniffi.examples.arithmeticpm.add
import io.gitlab.trixnity.uniffi.examples.arithmeticpm.sub
import io.gitlab.trixnity.uniffi.examples.todolist.TodoEntry
import io.gitlab.trixnity.uniffi.examples.todolist.TodoList

@Composable
fun ContentView(todoList: TodoList) {
    val stride = remember { 1UL }
    var clicked by remember { mutableStateOf(0UL) }
    var text by remember { mutableStateOf("") }

    Surface {
        Column(Modifier.fillMaxSize()) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(8.dp),
                horizontalArrangement = Arrangement.SpaceAround,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("$clicked")
                Spacer(Modifier.width(8.dp))
                TextField(
                    modifier = Modifier.fillMaxWidth(),
                    value = text,
                    onValueChange = { text = it },
                    placeholder = { Text("New Task") },
                    keyboardOptions = KeyboardOptions(imeAction = ImeAction.Send),
                    keyboardActions = KeyboardActions {
                        todoList.addEntry(TodoEntry("$clicked $text"))
                        text = ""
                        clicked = add(clicked, stride)
                    },
                    singleLine = true,
                )
            }

            LazyColumn(
                Modifier
                    .fillMaxSize()
                    .padding(8.dp),
            ) {
                items(todoList.getEntries()) { entry ->
                    ListItem(
                        headlineContent = { Text(entry.text) },
                        trailingContent = {
                            IconButton(
                                onClick = {
                                    todoList.clearItem(entry.text)
                                    clicked = sub(clicked, stride)
                                },
                            ) {
                                Icon(
                                    Icons.Outlined.Delete,
                                    contentDescription = "Delete ${entry.text}",
                                    tint = MaterialTheme.colorScheme.error
                                )
                            }
                        }
                    )
                }
            }
        }
    }
}

@Preview
@Composable
fun ContentViewPreview() {
    ContentView(TodoList())
}
